V0.17.0 bump (#1511)
PennyLaneAI/pennylane: .github/CHANGELOG.md

> # Release 0.17.0 (current release)

Reviewer: @antalszava I can't comment on it, but beware of accidental double-backticks :) I can see one down on line 44

![image](https://user-images.githubusercontent.com/2959003/128977718-c0b66708-9dc7-4469-bf40-d7817d375227.png)

Author: Updated! How come it's good to avoid them? :thinking: Would have thought both single and double-backticks work.

Reviewer: Oh - Markdown only supports single backticks for code :) Double backticks rendering correctly is accident of some renderers, and is not always guaranteed to work.

> <h4>Circuit optimization</h4>

Reviewer: ```suggestion
<h3>New features since the last release</h3>

<h4>Circuit optimization</h4>
```

# Release 0.17.0 (current release)

<h3>New features since the last release</h3>

<h4>Circuit optimization</h4>

* PennyLane can now perform quantum circuit optimization using the
  top-level transform `qml.compile`. The `compile` transform allows you
  to chain together sequences of tape and quantum function transforms
  into custom circuit optimization pipelines.
  [(#1475)](https://github.com/PennyLaneAI/pennylane/pull/1475)

  For example, take the following decorated quantum function:

  ```python
  dev = qml.device('default.qubit', wires=[0, 1, 2])

  @qml.qnode(dev)

>   @qml.transforms.compile()

Author: ```suggestion
  @qml.compile()
```

   2: ──H──╰C────────RZ(0.4)──RZ(-0.4)──RX(0.3)──╰CY──Y──┤
  ```

* A new quantum function transform has been added to push commuting
  single-qubit gates through controlled operations.

>   [(#1464)](https://github.com/PennyLaneAI/pennylane/pull/1464)

Reviewer: @antalszava @co9olguy I am wondering if having examples of every compilation transform might be information overload? An alternative is to keep the example for `qml.compile`, but replace the other examples with a condensed bullet list:

```markdown
  The following compilation transforms have been added and are also available
  to use, either independently, or within a `qml.compile` pipeline:

  * `commute_controlled`: push commuting single-qubit gates through controlled operations.
    [(#1464)](https://github.com/PennyLaneAI/pennylane/pull/1464)
    
  * `cancel_inverses`: removes adjacent pairs of operations that cancel out.
    [(#1455)](https://github.com/PennyLaneAI/pennylane/pull/1455)

  * `merge_rotations`: combines adjacent rotation gates of
    the same type into a single gate, including controlled rotations.
    [(#1455)](https://github.com/PennyLaneAI/pennylane/pull/1455)

  * `single_qubit_fusion`: acts on all sequences of
    single-qubit operations in a quantum function, and converts each
    sequence to a single `Rot` gate.
    [(#1458)](https://github.com/PennyLaneAI/pennylane/pull/1458)

  For more details on `qml.compile` and the available compilation transforms, see
  [the compilation documentation](https://pennylane.readthedocs.io/en/latest/code/qml_transforms.html#transforms-for-circuit-compilation).
```

Reviewer 2: yup I agree we can condense the story around compilation

  The `commute_controlled` transform works as follows:

  ```python
  def circuit(theta):
      qml.PauliX(wires=2)
      qml.S(wires=0)
      qml.CNOT(wires=[0, 1])
      qml.PhaseShift(theta/2, wires=0)
      qml.T(wires=0)
      qml.Toffoli(wires=[0, 1, 2])
      return qml.expval(qml.PauliZ(0))
  ```

  ```pycon
  >>> optimized_circuit = qml.transforms.commute_controlled(direction="right")(circuit)
  >>> dev = qml.device('default.qubit', wires=3)
  >>> qnode = qml.QNode(optimized_circuit, dev)
  >>> print(qml.draw(qnode)(0.5))
   0: ──╭C──╭C──S──Rϕ(0.25)──T──┤ ⟨Z⟩
   1: ──╰X──├C──────────────────┤
   2: ──────╰X──X───────────────┤

>   ```

Reviewer: For the compilation transforms, how do we feel about using decorator syntax instead?

```python
dev = qml.device('default.qubit', wires=3)

@qml.qnode(dev)
@qml.transforms.commute_controlled(direction="right")
def circuit(theta):
    qml.PauliX(wires=2)
    qml.S(wires=0)
    qml.CNOT(wires=[0, 1])
    qml.PhaseShift(theta/2, wires=0)
    qml.T(wires=0)
    qml.Toffoli(wires=[0, 1, 2])
    return qml.expval(qml.PauliZ(0))

>>> print(qml.draw(circuit)(0.5))
 0: ──╭C──╭C──S──Rϕ(0.25)──T──┤ ⟨Z⟩
 1: ──╰X──├C──────────────────┤
 2: ──────╰X──X───────────────┤
```

To me, it makes the examples easier to read, and slightly shorter (there is less duplication).

Reviewer 2: :+1:

Author: With the condensed bullet list of compilation transforms this example has been removed.

Reviewer: @antalszava there is the compilation example on lines 15-43 that can be updated to use decorator syntax :)

  [(#1458)](https://github.com/PennyLaneAI/pennylane/pull/1458)

  The `single_qubit_fusion` transform acts on all sequences of
  single-qubit operations in a quantum function, and converts each
  sequence to a single `Rot` gate. For example given the circuit:

  ```python
  def circuit(x, y, z):
      qml.Hadamard(wires=0)
      qml.PauliZ(wires=1)
      qml.RX(x, wires=0)
      qml.RY(y, wires=1)
      qml.CZ(wires=[1, 0])
      qml.T(wires=0)
      qml.SX(wires=0)
      qml.Rot(x, y, z, wires=1)
      qml.Rot(z, y, x, wires=1)
      return qml.expval(qml.PauliX(wires=0))
  ```

  ```pycon
  >>> optimized_circuit = qml.transforms.single_qubit_fusion()(circuit)
  >>> dev = qml.device('default.qubit', wires=2)
  >>> qnode = qml.QNode(optimized_circuit, dev)
  >>> print(qml.draw(qnode)(0.1, 0.2, 0.3))
   0: ──Rot(3.24, 1.57, 0)──╭Z──Rot(2.36, 1.57, -1.57)────┤ ⟨X⟩
   1: ──Rot(3.14, 0.2, 0)───╰C──Rot(0.406, 0.382, 0.406)──┤
  ```

<h4>Improved Hamiltonian simulations</h4>

* Added a sparse Hamiltonian observable and the functionality to support computing its expectation
  value. [(#1398)](https://github.com/PennyLaneAI/pennylane/pull/1398)

>   For example, the following QNode returns the expectation value of a sparse Hamiltonian:

Reviewer: I might be careful with this one:

- Only `default.qubit` is supported at the moment. However the entry makes it sound like it leads to faster execution in general.
- Only expvals are support, I believe(?)

Finally, I think the plan is to replace it with `qml.expval(qml.Hamiltonian(...))`, and simply use the fact that `qml.Hamiltonian` is _already_ a sparse representation of the Hamiltonian! So not sure how much we want to publicize this, since this change will likely be in the next release.

@mariaschuld, any thoughts here?

Reviewer: Perhaps it should be sufficient just to move this category 'lower' down the list of new features

Author: Sounds good! Moved it lower and added a note on `default.qubit`.

  ```python
  dev = qml.device("default.qubit", wires=2)

  @qml.qnode(dev, diff_method="parameter-shift")
  def circuit(param, H):
      qml.PauliX(0)
      qml.SingleExcitation(param, wires=[0, 1])
      return qml.expval(qml.SparseHamiltonian(H, [0, 1]))
  ```

  We can execute this QNode, passing in a sparse identity matrix:

  ```pycon
  >>> print(circuit([0.5], scipy.sparse.eye(4).tocoo()))
  0.9999999999999999
  ```

  The expectation value of the sparse Hamiltonian is computed directly, which leads to executions
  that are faster by orders of magnitude. Note that "parameter-shift" is the only differentiation
  method that is currently supported when the observable is a sparse Hamiltonian.

* VQE problems can now be intuitively set up by passing the Hamiltonian
  as an observable. [(#1474)](https://github.com/PennyLaneAI/pennylane/pull/1474)

  ``` python
  dev = qml.device("default.qubit", wires=2)
  H = qml.Hamiltonian([1., 2., 3.],  [qml.PauliZ(0), qml.PauliY(0), qml.PauliZ(1)])
  w = qml.init.strong_ent_layers_uniform(1, 2, seed=1967)

  @qml.qnode(dev)
  def circuit(w):
      qml.templates.StronglyEntanglingLayers(w, wires=range(2))

>       return qml.expval(H)

Reviewer: @mariaschuld just double checking here as well - I remember we had a conversation regarding whether this should be advertised now or in v0.18

  ```

  ```pycon
  >>> print(circuit(w))
  -1.5133943637878295
  >>> print(qml.grad(circuit)(w))
  [[[-8.32667268e-17  1.39122955e+00 -9.12462052e-02]
  [ 1.02348685e-16 -7.77143238e-01 -1.74708049e-01]]]
  ```

  Note that other measurement types like `var(H)` or `sample(H)`, as well
  as multiple expectations like `expval(H1), expval(H2)` are not supported.

* Added functionality to compute the sparse matrix representation of a `qml.Hamiltonian` object.
  [(#1394)](https://github.com/PennyLaneAI/pennylane/pull/1394)

> <h4> QNodes are more convenient</h4>

Reviewer: Trying to think of an alternative here that might be more descriptive 🤔

Author: Maybe "More convenient QNode usage"? I think both items here serve the purpose of user convenience. Therefore that might be a point worth having in the section name.

Reviewer: "QNodes are even more powerful"?

> * Added functionality to `qml.sample()` to extract samples from the basis states of
>   the device (currently only for qubit devices). Additionally, `wires` can be
>   specified to only return samples from those wires.
>   [(#1441)](https://github.com/PennyLaneAI/pennylane/pull/1441)

Reviewer: ```suggestion
* Computational basis samples directly from the underlying device can
  now be returned directly from QNodes via `qml.sample()`.
  [(#1441)](https://github.com/PennyLaneAI/pennylane/pull/1441)
```

Reviewer: This reminds me: there are several demos we should update to use this :) E.g., the [QAOA demo](https://pennylane.ai/qml/demos/tutorial_qaoa_maxcut.html) comes to mind (I think there are others).

Currently, existing demos use two separate tricks to get the computational basis samples:

1. They use `qml.Hermitian`:

   ```python
   return qml.sample(qml.Hermitian(np.diag(2 ** np.arange(num_wires), wires=wires))

   res = qnode(params) # returns list of state samples in base-10
   output_samples = convert_to_binary(res) # need to convert to binary representation
   ```

2. Or, they use

   ```python
    return [qml.sample(qml.PauliZ(w)) for w in wires]

    res = qnode(params) # returns state samples using -1, 1
    output_samples = (res + 1) / 2 # convert to using 0, 1
    ```

Both of these approaches add overhead/complexity (_especially_ number 1!) and so should be updated. It also helps propagate to our users this new functionality.

Author: Made a note!

  @qml.qnode(dev)
  def circuit_1():
      qml.Hadamard(wires=0)
      qml.Hadamard(wires=1)
      return qml.sample()

  @qml.qnode(dev)
  def circuit_2():
      qml.Hadamard(wires=0)
      qml.Hadamard(wires=1)
      return qml.sample(wires=[0,2])    # no observable provided and wires specified
  ```

  ```pycon
  >>> print(circuit_1())
  [[1, 0, 0],
   [1, 1, 0],
   [1, 0, 0],
   [0, 0, 0],
   [0, 1, 0]]

  >>> print(circuit_2())
  [[1, 0],
   [1, 0],
   [1, 0],
   [0, 0],
   [0, 0]]

  >>> print(qml.draw(circuit_2))
   0: ──H──╭┤ Sample[basis]
   1: ──H──│┤
   2: ──H──╰┤ Sample[basis]
  ```

* The new ``qml.apply`` function can be used to add operations that might have
  already been instantiated elsewhere to the QNode and other queuing contexts:
  [(#1433)](https://github.com/PennyLaneAI/pennylane/pull/1433)

  ```python
  op = qml.RX(0.4, wires=0)
  dev = qml.device("default.qubit", wires=2)

  @qml.qnode(dev)
  def circuit(x):
      qml.RY(x, wires=0)
      qml.apply(op)
      return qml.expval(qml.PauliZ(0))
  ```

  ```pycon
  >>> print(qml.draw(circuit)(0.6))
  0: ──RY(0.6)──RX(0.4)──┤ ⟨Z⟩
  ```

  Previously instantiated measurements can also be applied to QNodes.

> <h4>Resource Devicer Tracker</h4>

Reviewer: There is a rendering issue with the device tracker below:

![image](https://user-images.githubusercontent.com/2959003/128978956-41f1ed5a-0e88-430c-bc3c-a02410146b59.png)

Author: Good catch! Changed it.

Reviewer 2: ```suggestion
<h4>Resource Device Tracker</h4>
```

  Total shots:  100
  Total shots:  200
  Total shots:  300
  Total shots:  300
  ```

<h4>Containerization support</h4>

* Docker support for building PennyLane with support for all interfaces (TensorFlow,
  Torch, and Jax), as well as device plugins and QChem, for GPUs and CPUs, has been added.
  [(#1391)](https://github.com/PennyLaneAI/pennylane/pull/1391)

>   The build process using Docker and Makefile works as follows:

Reviewer: ```suggestion
  The build process using Docker and `make` requires that the repository source code
  is cloned or downloaded from GitHub, and provides the following options:
```

Reviewer: @antalszava I just noticed something; the following instructions are also provided in the README, however they shouldn't be there - we made a concerted effort to simplify the README a couple of releases ago!

I recommend the following:

- Remove the docker section from the README
- Instead, add it to the developer guide as a new card here: https://pennylane.readthedocs.io/en/latest/development/guide.html
- Add a short link in the readme
- Add a short link here in the changelog, and remove the full list of commands

Author: Good point! Removed the docker section from the README. It also seems like we could add the description about Docker to the installation parts of the development guide and the user installation section. Let me open a separate PR.

  Total shots:  300
  ```

<h4>Containerization support</h4>

* Docker support for building PennyLane with support for all interfaces (TensorFlow,
  Torch, and Jax), as well as device plugins and QChem, for GPUs and CPUs, has been added.
  [(#1372)](https://github.com/PennyLaneAI/pennylane/issues/1372)

  The build process using Docker and Makefile works as follows:

>   Building a core PennyLane image:

Reviewer: Seems like these should be bullet points?

Author: Good catch! Updated

   [ 0.69916862  0.34072424  0.69202359]]
  ```

> <h4>Extended operations and templates</h4>

Reviewer: ```suggestion
<h4>Even more new operations and templates</h4>
```

Reviewer: > A decomposition has been added to QubitUnitary that makes the single-qubit case fully differentiable in all interfaces. 

This one is really cool! Should we move it higher?

Author: Would you suggest moving it higher within the "Even more new operations and templates" section of outside of it even?

Reviewer: I was thinking more moving it higher inside the "new operations" section :)

This one is something a lot of users have asked about (training qubit unitary), so is a nice one to make more visible

* Change the order of the covariance matrix and the vector of means internally
  in `default.gaussian`. [(#1331)](https://github.com/PennyLaneAI/pennylane/pull/1331)

> * Added the `id` attribute to templates, which was missing from
>   PR [(#1377)](https://github.com/PennyLaneAI/pennylane/pull/1377).
>   [(#1438)](https://github.com/PennyLaneAI/pennylane/pull/1438)

Reviewer 2: ```suggestion
* Added the `id` attribute to templates). [(#1438)](https://github.com/PennyLaneAI/pennylane/pull/1438)
```

> 

Author: Thanks! Updated.
